torch/testing/_internal/common_subclass.py

```diff
@@ -0,0 +1,261 @@
+
+    def __init__(self, size, values, indices, requires_grad=False):
+        self.values = values
+        self.indices = indices
+
+    def __repr__(self):
+        return super().__repr__(tensor_contents=f"values={self.values}, indices={self.indices}")
+
+    def sparse_to_dense(self):
+        res = torch.zeros(self.size(), dtype=self.values.dtype)
+        res[self.indices.unbind(1)] = self.values
+        return res
+
+    @staticmethod
+    def from_dense(t):
+        indices = t.nonzero()
+        values = t[indices.unbind(1)]
+        return SparseTensor(t.size(), values, indices)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        func_name = f"{func.__module__}.{func.__name__}"
+
+        res = cls._try_call_special_impl(func_name, args, kwargs)
+        if res is not NotImplemented:
+            return res
+
+        # Otherwise, use a default implementation that construct dense
+        # tensors and use that to compute values
+        def unwrap(e):
+            return e.sparse_to_dense() if isinstance(e, SparseTensor) else e
+
+        # Wrap back all Tensors into our custom class
+        def wrap(e):
+            # Check for zeros and use that to get indices
+            return SparseTensor.from_dense(e) if isinstance(e, torch.Tensor) else e
+
+        # Convolution is being reworked to avoid this. But it didn't land yet
+        if func_name == "torch._ops.aten.convolution_overrideable":
+            # Go back to plain convolution to dispatch to the right device now
+            # that the input is a plain cpu tensor
+            func = torch.ops.aten.convolution
+        if func_name == "torch._ops.aten.convolution_backward_overrideable":
+            # There is no generic conv backward (yet) but in colab we always
+            # use mkldnn so that's an easy workaround
+            func = torch.ops.aten.mkldnn_convolution_backward
+            # Re-order the arguments to match the new signature
+            assert args[6] is False
+            assert args[7] == [0, 0]
+            # Undefined Tensors are not properly handled by the pybind layer?
+            grad_mask = [True, True, True]
+            args = (args[1], args[0], args[2], args[4], args[3], args[5], args[8], grad_mask)
+
+
+        rs = tree_map(wrap, func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs or {})))
+        return rs
+
+    # To show how things happen later
+    def __rmul__(self, other):
+        return super().__rmul__(other)
+
+    _SPECIAL_IMPLS = {}
+
+    @classmethod
+    def _try_call_special_impl(cls, func, args, kwargs):
+        if func not in cls._SPECIAL_IMPLS:
+            return NotImplemented
+        return cls._SPECIAL_IMPLS[func](args, kwargs)
+
+
+# Example non-wrapper subclass that stores extra state.
+class NonWrapperTensor(torch.Tensor):
+    def __new__(cls, data):
+        t = data.clone()
+        t.__class__ = cls
```

Comment:
We should discourage people from doing this, since `__torch_dispatch__` will not work for subclasses initialized in this way.

Reply:
cc @albanD

Reply:
Related: is there a canonical example for a non-wrapper subclass with extra state? I didn't see one so I hacked this together - should this sort of thing be discouraged entirely? Otherwise, what should creation look like?

Reply:
Agreed!

Btw I think the users should never implement the __new__ by hand!